controller/engine_controller.go

```diff
@@ -2215,3 +2238,75 @@ func shouldProceedToWaitAndRebuild(e *longhorn.Engine, replicaName, originalRepl
 
 	return true
 }
+
+// needStatusUpdate checks whether we should update the engine status and whether that update should be rate limited. We
+// return:
+// - false, false if no fields change
+// - true, false if any field besides the size of the volume-head snapshot changes
+// - true, false if the change in size of the volume-head snapshot exceeds a threshold
+// - true, true if the change in size of the volume-head does not exceed a threshold
+func (m *EngineMonitor) needStatusUpdate(existing, new *longhorn.Engine) (needStatusUpdate, rateLimited bool) {
+	if needStatusUpdateBesidesSize(&existing.Status, &new.Status) {
+		return true, false
+	}
+
+	// Existence was already verified in needStatusUpdateBesidesSize.
+	existingSnapshot := existing.Status.Snapshots[etypes.VolumeHeadName]
+	newSnapshot := new.Status.Snapshots[etypes.VolumeHeadName]
+
+	// Now, compare only the volume-head sizes.
+	var existingSizeInt, newSizeInt int64
+	var err error
+	if existingSizeInt, err = strconv.ParseInt(existingSnapshot.Size, 10, 64); err != nil {
+		m.logger.WithError(err).Warnf("Failed to parse %s size %v", etypes.VolumeHeadName, existingSnapshot.Size)
+		return false, false
+	}
+	if newSizeInt, err = strconv.ParseInt(newSnapshot.Size, 10, 64); err != nil {
+		m.logger.WithError(err).Warnf("Failed to parse %s size %v", etypes.VolumeHeadName, newSnapshot.Size)
+		return false, false
+	}
+	return needSizeUpdate(existingSizeInt, newSizeInt, new.Spec.VolumeSize)
+}
+
+// needStatusUpdateBesidesSize does half of the work of needStatusUpdate. It is broken out because only one caller
+// should consider the size of the volume-head snapshot when deciding whether to update. All other callers should not
+// consider a change in volume-head snapshot size a reason to update. If they do, they will circumvent rate limiting.
+func needStatusUpdateBesidesSize(existing, new *longhorn.EngineStatus) bool {
+	// If we don't have a volume-head snapshot in new and old status, just compare statuses directly.
+	existingSnapshot, existingSnapshotOK := existing.Snapshots[etypes.VolumeHeadName]
+	newSnapshot, newSnapshotOK := new.Snapshots[etypes.VolumeHeadName]
+	if !existingSnapshotOK || !newSnapshotOK {
+		return !reflect.DeepEqual(existing, new)
+	}
+
+	// Otherwise, compare without the size of volume-head.
+	existingSize := existingSnapshot.Size
+	existingSnapshot.Size = newSnapshot.Size
+	needUpdate := !reflect.DeepEqual(existing, new)
+	existingSnapshot.Size = existingSize
```

Comment:
A safer and much more laborious way would be to make copies of new and existing snapshot and zero out the respective Size fields for the DeepEqual check.  That has the advantage of not doctoring the input arguments, which one would expect to be `const` in some sense.  We just hope that nobody makes a copy or refers to the Size field for the brief moment that it is being hijacked.  You probably considered that and decided that the risk was tiny and the overhead of the full copy just for comparison was pretty heavyweight for such a mild exposure.

Reply:
I agree with your analysis here, and you are correct about my thought process. Since `existingEngine` itself is a deep copy created just for this refresh and has no practical way of being operated simultaneously by another actor, the exposure is small. I didn't want to deep copy the entire `EngineStatus` struct (I think there's no way around comparing the whole thing) additionally every five seconds to mitigate the risk.